File: python/src/azoth/eos/reference/solid_fugacity.py
"""``eos.solid_fugacity`` - a pure solid's fugacity coefficient, from tabulated properties.

```text
phi_solid = phi_liq(T, P) exp( -dH_fus/(R T) (1 - T/T_tp)
                             + dCp_SL/(R T) (T_tp - T)
                             - dCp_SL/R ln(T_tp/T)
                             - dV_sl (P_bar - 1)/(R T) )
```

Spec: ``specs/calcs/eos/solid_fugacity.toml``, which carries the heat-capacity difference's
two routes and the measurement that pins the whole expression.

NeqSim's ``ComponentSolid.fugcoef2``. **The reference is a *liquid*** - ``fugcoef2``
initialises it ``PhaseType.LIQUID``, where the class's other entry (``fugcoef``) uses a gas at
the component's solid vapour pressure and is a different model - and it is the host's own
class, so ``eos`` is an input.
"""

from __future__ import annotations

import math

from azoth._registry_gen import spec as _spec_for
from azoth.core.errors import InvalidInputError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import SolidFugacityResult
from azoth.core.units import Q, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.cubic import PR, SRK
from azoth.eos.mixture import Component, Mixture
from azoth.eos.reference._mixture_state import phase_state, reduced_parameters

CALC_ID = "eos.solid_fugacity"

#: NeqSim's ``R``, which ``Component`` states as a literal.
R = 8.3144621

#: One atmosphere in bar: the pressure the volume term is referred to in NeqSim's own code.
REFERENCE_PRESSURE_BAR = 1.0


def solid_fugacity(
    heat_of_fusion: Q,
    triple_point_temperature: Q,
    delta_cp_sl: Q,
    delta_solid_volume: Q,
    tc: Q,
    pc: Q,
    omega: float,
    T: Q,
    P: Q,
    eos: str = "srk",
) -> SolidFugacityResult:
    """A pure solid's fugacity coefficient at a state.

    Args:
        heat_of_fusion: the component's heat of fusion.
        triple_point_temperature: its triple-point temperature.
        delta_cp_sl: the solid-liquid heat-capacity difference **at the triple point**, which
            the caller states: NeqSim takes it from the table and overrides it to ``37.12``
            for water.
        delta_solid_volume: the solid's molar volume less the liquid's, at the triple point.
        tc: the component's critical temperature.
        pc: its critical pressure.
        omega: its acentric factor.
        T: absolute temperature.
        P: absolute pressure.
        eos: the cubic the fluid runs, which is also the one the reference liquid is built
            from.

    Raises:
        InvalidInputError: if ``eos`` is not a cubic this reaches.
        OutOfRangeError: if any of the constants or the state is not positive.
    """
    spec = _spec_for(CALC_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []
    fusion_si = input_to_si(spec, "heat_of_fusion", heat_of_fusion)
    ttp_si = input_to_si(spec, "triple_point_temperature", triple_point_temperature)
    dcp_si = input_to_si(spec, "delta_cp_sl", delta_cp_sl)
    dvol_si = input_to_si(spec, "delta_solid_volume", delta_solid_volume)
    t_si = input_to_si(spec, "T", T)
    p_si = input_to_si(spec, "P", P)
    apply_checks(
        checks.on_input,
        {
            "heat_of_fusion": fusion_si,
            "triple_point_temperature": ttp_si,
            "T": t_si,
            "P": p_si,
        }.get,
        warnings,
    )

    if eos not in ("pr", "srk"):
        raise InvalidInputError(
            "eos",
            f"`{eos}` is not a cubic this reaches: NeqSim builds the reference liquid from the "
            "*host phase's* class, and only `pr` and `srk` are ported",
        )

    reference = Mixture(
        components=(Component(Tc=tc, Pc=pc, omega=omega),),
        cubic=PR if eos == "pr" else SRK,
        alpha="pr" if eos == "pr" else "srk",
    )
    reduced = reduced_parameters(reference, t_si, p_si)
    state = phase_state(reduced, reference.kij, [1.0], liquid=True)
    phi_liquid = math.exp(state.ln_phi[0])

    pressure_bar = p_si / 1.0e5
    fusion = -fusion_si / (R * t_si) * (1.0 - t_si / ttp_si)
    heat_capacity = dcp_si / (R * t_si) * (ttp_si - t_si) - dcp_si / R * math.log(ttp_si / t_si)
    volume = -dvol_si * (pressure_bar - REFERENCE_PRESSURE_BAR) / (R * t_si)

    return SolidFugacityResult(
        fugacity_coefficient=phi_liquid * math.exp(fusion + heat_capacity + volume),
        warnings=tuple(warnings),
    )


__all__ = ["solid_fugacity"]

File: crates/azoth-eos/src/solid_fugacity.rs
//! `eos.solid_fugacity` - a pure solid's fugacity coefficient, from tabulated properties.
//!
//! ```text
//! phi_solid = phi_liq(T, P) exp( -dH_fus/(R T) (1 - T/T_tp)
//!                              + dCp_SL/(R T) (T_tp - T)
//!                              - dCp_SL/R ln(T_tp/T)
//!                              - dV_sl (P_bar - 1)/(R T) )
//! ```
//!
//! Spec: `specs/calcs/eos/solid_fugacity.toml`, which carries the three input routes for the
//! heat-capacity difference and the measurement that pins the whole expression.
//!
//! NeqSim's `ComponentSolid.fugcoef2`. **The reference is a *liquid*** - `fugcoef2`
//! initialises it `PhaseType.LIQUID`, where the class's other entry (`fugcoef`) uses a gas at
//! the component's solid vapour pressure and is a different model - and it is the host's own
//! class, so `eos` is an input and not a label.

use azoth_core::units::{Pressure, ThermodynamicTemperature};
use azoth_core::{AzothError, Result, apply_checks};

use crate::Alpha;
use crate::Cubic;
use crate::mixture::{Component, Mixture, RootSide};
use crate::results::SolidFugacityResult;
use crate::spec_gen;

/// NeqSim's `R`, which `Component` states as a literal.
pub const R: f64 = 8.314_462_1;

/// One atmosphere in bar: the pressure the volume term is referred to in NeqSim's own code.
const REFERENCE_PRESSURE_BAR: f64 = 1.0;

/// A pure solid's fugacity coefficient at a state.
///
/// # Errors
/// * [`AzothError::InvalidInput`] if `eos` is not a cubic this reaches.
/// * [`AzothError::OutOfRange`] if any of the constants or the state is not positive.
#[allow(clippy::too_many_arguments, non_snake_case)]
pub fn solid_fugacity(
    heat_of_fusion: f64,
    triple_point_temperature: f64,
    delta_cp_sl: f64,
    delta_solid_volume: f64,
    tc: ThermodynamicTemperature,
    pc: Pressure,
    omega: f64,
    T: ThermodynamicTemperature,
    P: Pressure,
    eos: &str,
) -> Result<SolidFugacityResult> {
    let spec = &spec_gen::SOLID_FUGACITY_SPEC;
    let mut warnings = Vec::new();
    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "heat_of_fusion" => Some(heat_of_fusion),
            "triple_point_temperature" => Some(triple_point_temperature),
            "T" => Some(T.value),
            "P" => Some(P.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let cubic = match eos {
        "pr" => Cubic::Pr,
        "srk" => Cubic::Srk,
        other => {
            return Err(AzothError::invalid_input(
                "eos",
                format!(
                    "`{other}` is not a cubic this reaches: NeqSim builds the reference liquid \
                     from the *host phase's* class, and only `pr` and `srk` are ported"
                ),
            ));
        }
    };

    // **The reference liquid**: this component alone, on the fluid's own cubic, on the liquid
    // root.
    let alpha = match cubic {
        Cubic::Pr => Alpha::Pr,
        _ => Alpha::Srk,
    };
    let reference = Mixture::new(vec![Component::new(tc, pc, omega)?], vec![0.0])?
        .with_cubic(cubic)
        .with_alpha(alpha);
    let reduced = reference.reduced_parameters(T, P)?;
    let state = reference.phase_state(&reduced, &[1.0], RootSide::Liquid)?;
    let phi_liquid = state.ln_phi[0].exp();

    let pressure_bar = P.value / 1.0e5;
    let fusion = -heat_of_fusion / (R * T.value) * (1.0 - T.value / triple_point_temperature);
    let heat_capacity = delta_cp_sl / (R * T.value) * (triple_point_temperature - T.value)
        - delta_cp_sl / R * (triple_point_temperature / T.value).ln();
    let volume = -delta_solid_volume * (pressure_bar - REFERENCE_PRESSURE_BAR) / (R * T.value);

    Ok(SolidFugacityResult {
        fugacity_coefficient: phi_liquid * (fusion + heat_capacity + volume).exp(),
        warnings,
    })
}
